src/neorados/cls/log.cc

```diff
@@ -0,0 +1,132 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 IBM
+ *
+ * See file COPYING for license information.
+ *
+ */
+
+#include "log.h"
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "include/buffer.h"
+#include "common/ceph_time.h"
+
+#include "include/neorados/RADOS.hpp"
+
+
+#include "cls/log/cls_log_types.h"
+#include "cls/log/cls_log_ops.h"
+
+namespace neorados::cls::log {
+using boost::system::error_code;
+namespace buffer = ceph::buffer;
+
+void add(neorados::WriteOp& op, std::vector<entry> entries)
+{
+  buffer::list in;
+  ::cls::log::ops::add_op call;
+  call.entries = std::move(entries);
+  encode(call, in);
+  op.exec("log", "add", in);
+}
+
+void add(neorados::WriteOp& op, entry e)
+{
+  bufferlist in;
+  ::cls::log::ops::add_op call;
+  call.entries.push_back(std::move(e));
+  encode(call, in);
+  op.exec("log", "add", in);
+}
+
+void add(neorados::WriteOp& op, ceph::real_time timestamp,
+	 std::string section, std::string name,
+	 buffer::list&& bl)
+{
+  bufferlist in;
+  ::cls::log::ops::add_op call;
+  call.entries.emplace_back(timestamp, std::move(section),
+			    std::move(name), std::move(bl));
+  encode(call, in);
+  op.exec("log", "add", in);
+}
+
+void list(ReadOp& op, ceph::real_time from, ceph::real_time to,
+	  std::optional<std::string> in_marker, std::span<entry>& entries,
+	  std::optional<std::string>* const out_marker)
+{
+  bufferlist inbl;
+  ::cls::log::ops::list_op call;
+  call.from_time = from;
+  call.to_time = to;
+  call.marker = std::move(in_marker).value_or("");
+  call.max_entries = entries.size();
+
+  encode(call, inbl);
+  op.exec("log", "list", inbl,
+          [&entries, out_marker](error_code ec, const buffer::list& bl) {
+            ::cls::log::ops::list_ret ret;
+            if (!ec) {
+	      auto iter = bl.cbegin();
+	      decode(ret, iter);
+	      entries = entries.first(ret.entries.size());
+	      std::move(ret.entries.begin(), ret.entries.end(),
+			entries.begin());
+	      if (out_marker) {
+		*out_marker = (ret.truncated ?
+			       std::move(ret.marker) :
+			       std::optional<std::string>{});
```

Comment:
I prefer it mostly so we can test marker for truth/falsehood.

But `nullopt` doesn't work in this context since it violates the rule that both alternatives of a ternary expression have to be the same type. It comes as `nullopt_t` rather than `std::optional<whatever>`

Reply:
oops, i misread that as initializing an optional with an empty string, rather than default-constructing an empty optional